File: src/main/java/com/antwerkz/ophelia/dao/OpheliaDao.java
package com.antwerkz.ophelia.dao;

import com.antwerkz.ophelia.models.MongoModel;
import org.bson.types.ObjectId;
import org.mongodb.morphia.Datastore;

public class OpheliaDao<T> {
  private Datastore ds;

  private Class<T> clazz;

  public OpheliaDao(Datastore ds, Class<T> clazz) {
    this.ds = ds;
    this.clazz = clazz;
  }

  public Datastore getDs() {
    return ds;
  }

  public T find(final ObjectId id) {
    return ds.createQuery(clazz).field("_id").equal(id).get();
  }

  public void save(MongoModel model) {
    ds.save(model);
  }

  public void delete(MongoModel model) {
    ds.delete(model);
  }
}

File: src/main/kotlin/com/antwerkz/ophelia/dao/OpheliaDao.kt
package com.antwerkz.ophelia.dao

import com.antwerkz.ophelia.models.MongoModel
import org.bson.types.ObjectId
import org.mongodb.morphia.Datastore

public open class OpheliaDao<T>(public val ds: Datastore, private val clazz: Class<T>) {

    public fun find(id: ObjectId): T {
        return ds.createQuery<T>(clazz).field("_id").equal(id).get()
    }

    public fun save(model: MongoModel<T>) {
        ds.save<MongoModel<T>>(model)
    }

    public fun delete(model: MongoModel<Any>) {
        ds.delete<MongoModel<Any>>(model)
    }
}
